docs/book/middleware.md

```diff
@@ -5,41 +5,81 @@ for your message data, you have to convert this data to an array before you can
 > Note: The middleware uses an array for the message data
 
 ## CommandMiddleware
-The `CommandMiddleware` dispatches the message data to the command bus system. This middleware needs an request attribute 
-(`$request->getAttribute(\Prooph\HttpMiddleware\CommandMiddleware::NAME_ATTRIBUTE)`) called `prooph_command_name`. 
-This name is used for the `\Prooph\Common\Messaging\MessageFactory` to create the `\Prooph\Common\Messaging\Message` 
-object. The data for the command is extracted from the body of the request (`$request->getParsedBody()`) and must be an 
+The `CommandMiddleware` dispatches the message data to the command bus system. This middleware needs an request attribute
+(`$request->getAttribute(\Prooph\HttpMiddleware\CommandMiddleware::NAME_ATTRIBUTE)`) called `prooph_command_name`.
+This name is used for the `\Prooph\Common\Messaging\MessageFactory` to create the `\Prooph\Common\Messaging\Message`
+object. The data for the command is extracted from the body of the request (`$request->getParsedBody()`) and must be an
 array.
 
 ## QueryMiddleware
-The `QueryMiddleware` dispatches the message data to the query bus system. This middleware needs an request attribute 
-(`$request->getAttribute(\Prooph\HttpMiddleware\QueryMiddleware::NAME_ATTRIBUTE)`) called `prooph_query_name`. 
-This name is used for the `\Prooph\Common\Messaging\MessageFactory` to create the `\Prooph\Common\Messaging\Message` 
-object. The data for the query is extracted from query params of the request (`$request->getQueryParams()`) and must be an 
-array.
+The `QueryMiddleware` dispatches the message data to the query bus system. Unlike the other middleware, the
+`QueryMiddleware` supports sending multiple messages using a *POST* request.
+
+### Using GET
+
+Using *GET*, the `QueryMiddleware` only supports a single message. The `prooph_query_name` attribute must be present as
+an `attribute` on the request. Any additional query parameters will be added to the body of the message as well. An
+example URL template would be
+
+`GET /query/{prooph_query_name}`
+
+The resulting request might be
+
+`GET /query/users?name=John`
+
+And the parsed message array would end up as
+
+```php
+[
+    'prooph_query_name' => 'users',
+    'name' => 'John'
+]
+```
+
+
+### Using POST
+
+With a *POST* request, it is expected that the parsed body contain an array of messages. Each object of the array must
+contain the `prooph_query_name` property. All other information will be used when populating the the message object. An
+example JSON request might look something like
 
-There is a special behaviour implemented. If you send a *POST* HTTP request, then the parsed body data (`$request->getParsedBody()`) 
-will be added to the payload under the key `data`. `data` is a reserved key if you use a *POST* HTTP request. However, it's
-not recommended to use a *POST* HTTP request here. Use it only if you know what you do.
+`POST /query`
+
+```json
```

Comment:
Maybe we should use another structure for such type of request.
- If the whole request is an array, there is no way to provide some infos to other parts of application code.
- What would happen if the same `prooph_query_name` is defined more than once? How can i ensure the correct response order?

I would prefer something like that:
```json
{
    "prooph_queries": [
        {
                "message_name": "get-users",
                "payload": {
                     "filter": [12]
                }
         }
    ]
}
```

AND / OR the named version

```json
{
    "prooph_queries": {
        "query_result_indicator": {
                "message_name": "get-users",
                "payload": {
                     "filter": [12]
                }
         }
    }
}
```

Reply:
I think the idea of having some sort of "query identifier" is good. I have some further thoughts on it as well, and I will try to put some time to making sure my thoughts actually pan out tomorrow.

Reply:
That's a good idea, maybe the result should more look like this then:

```
$promises['query_result_indicator'] = $this->queryBus->dispatch($query);
```

This way I don't need to make the finders depending on returning some subarray.

I hope it's clear what I mean, if not please hit me again :)